packages/expo-sqlite/android/src/main/java/expo/modules/sqlite/SQLiteModule.kt

```diff
@@ -0,0 +1,326 @@
+        }
+        if (!dir.mkdirs()) {
+          var additionalErrorMessage = ""
+          if (dir.exists()) {
+            additionalErrorMessage = "Path already points to a non-normal file."
+          }
+          if (dir.parentFile == null) {
+            additionalErrorMessage = "Parent directory is null."
+          }
+          throw IOException("Couldn't create directory '$dir'. $additionalErrorMessage")
+        }
+      }
+      return dir
+    }
+
+    private fun pluginResultsToPrimitiveData(results: Array<SQLitePluginResult?>): List<Any> {
+      val list: MutableList<Any> = ArrayList()
+      for (i in results.indices) {
+        val result = results[i]
+        val arr = convertPluginResultToArray(result)
+        list.add(arr)
+      }
+      return list
+    }
+
+    private fun convertPluginResultToArray(result: SQLitePluginResult?): List<Any?> {
+      val data: MutableList<Any?> = ArrayList()
+      data.add(result!!.error?.message)
+      data.add(result.insertId.toInt())
+      data.add(result.rowsAffected)
+
+      // column names
+      val columnNames: MutableList<String?> = ArrayList()
+      for (i in result.columns.indices) {
+        columnNames.add(result.columns[i])
+      }
+      data.add(columnNames)
+
+      // rows
+      val rows: MutableList<Any> = ArrayList()
+      for (i in result.rows.indices) {
+        val values = result.rows[i] ?: continue
+
+        // row content
+        val rowContent: MutableList<Any?> = ArrayList()
+        for (j in values.indices) {
+          val value = values[j]
+          if (value == null) {
+            rowContent.add(null)
+          } else if (value is String) {
+            rowContent.add(value as String?)
+          } else if (value is Boolean) {
+            rowContent.add(value as Boolean?)
+          } else {
+            val v = value as Number
+            rowContent.add(v.toDouble())
+          }
+        }
+        rows.add(rowContent)
+      }
+      data.add(rows)
+      return data
+    }
+
+    private fun isPragma(str: String): Boolean {
+      return startsWithCaseInsensitive(str, "pragma")
+    }
+
+    private fun isPragmaReadOnly(str: String): Boolean {
+      if (!isPragma(str)) {
+        return false
+      }
+      return !str.contains('=')
+    }
+
+    private fun isSelect(str: String): Boolean {
+      return startsWithCaseInsensitive(str, "select") || isPragmaReadOnly(str)
+    }
+
+    private fun isInsert(str: String): Boolean {
+      return startsWithCaseInsensitive(str, "insert")
+    }
+
+    private fun isUpdate(str: String): Boolean {
+      return startsWithCaseInsensitive(str, "update")
+    }
+
+    private fun isDelete(str: String): Boolean {
+      return startsWithCaseInsensitive(str, "delete")
+    }
+
+    // identify an "insert"/"select" query more efficiently than with a Pattern
+    private fun startsWithCaseInsensitive(str: String, substr: String): Boolean {
```

Comment:
I think we can use the `substring` function instead of our own implementation. cc @bbarthec

Reply:
I've done a quick benchmark to check and this one seems faster (compared to trimStart + startsWith), I would keep it

Reply:
@ixf, what order of magnitude you're writing about (in terms of being faster)?
As long as we're staying in the same time complexity category simplification is the better choice. I'm quite opinionated in this matter, so let's go with `trimStart + startsWith (+ignoreCase flag)`